src/utilities/schemaComparisons.js

```diff
@@ -0,0 +1,244 @@
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+import {
+  getNamedType,
+  GraphQLEnumType,
+  GraphQLInputObjectType,
+  GraphQLInterfaceType,
+  GraphQLObjectType,
+  GraphQLUnionType,
+} from '../type/definition';
+import {
+  GraphQLSchema,
+} from '../type/schema';
+
+export const BreakingChangeType = {
+  FIELD_CHANGED_KIND: 'FIELD_CHANGED_KIND',
+  FIELD_REMOVED: 'FIELD_REMOVED',
+  TYPE_CHANGED_KIND: 'TYPE_CHANGED_KIND',
+  TYPE_REMOVED: 'TYPE_REMOVED',
+  TYPE_REMOVED_FROM_UNION: 'TYPE_REMOVED_FROM_UNION',
+  VALUE_REMOVED_FROM_ENUM: 'VALUE_REMOVED_FROM_ENUM',
+};
+
+export type BreakingChange = {
+  type: $Keys<typeof BreakingChangeType>;
+  description: string;
+};
+
+/**
+ * Given two schemas, returns an Array containing descriptions of all the types
+ * of breaking changes covered by the other functions down below.
+ */
+export function findBreakingChanges(
+  oldSchema: GraphQLSchema,
+  newSchema: GraphQLSchema
+): Array<BreakingChange> {
+  return [
+    ...findRemovedTypes(oldSchema, newSchema),
+    ...findTypesThatChangedKind(oldSchema, newSchema),
+    ...findFieldsThatChangedType(oldSchema, newSchema),
+    ...findTypesRemovedFromUnions(oldSchema, newSchema),
+    ...findValuesRemovedFromEnums(oldSchema, newSchema)
+  ];
+}
+
+/**
+ * Given two schemas, returns an Array containing descriptions of any breaking
+ * changes in the newSchema related to removing an entire type.
+ */
+export function findRemovedTypes(
+  oldSchema: GraphQLSchema,
+  newSchema: GraphQLSchema
+): Array<BreakingChange> {
+  const oldTypes = Object.keys(oldSchema.getTypeMap());
+  const newTypes = new Set(Object.keys(newSchema.getTypeMap()));
+  return oldTypes.filter(typeName => !newTypes.has(typeName)).map(
+    type => ({
+      type: BreakingChangeType.TYPE_REMOVED,
+      description: `${type} was removed`,
+    })
+  );
+}
+
+/**
+ * Given two schemas, returns an Array containing descriptions of any breaking
+ * changes in the newSchema related to changing the type of a type.
+ */
+export function findTypesThatChangedKind(
+ oldSchema: GraphQLSchema,
+ newSchema: GraphQLSchema
+): Array<BreakingChange> {
+  const oldTypeMap = oldSchema.getTypeMap();
+  const newTypeMap = newSchema.getTypeMap();
+
+  const breakingChanges = [];
+  Object.keys(oldTypeMap).forEach(typeName => {
+    if (!newTypeMap[typeName]) {
+      return;
+    }
+    const oldType = oldTypeMap[typeName];
+    const newType = newTypeMap[typeName];
+    if (!(oldType instanceof newType.constructor)) {
+      breakingChanges.push(
+        {
+          type: BreakingChangeType.TYPE_CHANGED_KIND,
+          description: `${typeName} changed from ` +
+            `${prettifyTypeKind(oldType.constructor.name)} type to ` +
+            `${prettifyTypeKind(newType.constructor.name)} type`,
+        }
+      );
+    }
+  });
+  return breakingChanges;
+}
+
+function prettifyTypeKind(typeKind: string): string {
```

Comment:
I think just adding a `prettyName` static property to each graphql type class would comply the open-closed principle even better. It could be accessed as `oldType.constructor.prettyName`.